graphql_compiler/compiler/ir_lowering_match.py

```diff
@@ -578,6 +593,194 @@ def prune_output_blocks_in_compound_match_query(compound_match_query):
+    if not isinstance(predicate_list, list):
+        raise AssertionError(u'Expected `list`, Received {}.'.format(predicate_list))
+    if not predicate_list:
+        return None
+
+    if not isinstance(predicate_list[0], Expression):
+        raise AssertionError(u'Non-predicate object {} found in predicate_list'
+                             .format(predicate_list[0]))
+    if len(predicate_list) == 1:
+        return predicate_list[0]
+    else:
+        return BinaryComposition(u'&&',
+                                 _predicate_list_to_where_block(predicate_list[1:]),
+                                 predicate_list[0])
+
+
+def _collect_filters_to_first_location_in_match_traversal(match_traversal, location_to_predicates):
+    """Compose all filters for a specific location into its first occurence in given traversal."""
+    new_match_traversal = []
+    for match_step in match_traversal:
+        # Apply all filters for a location to the first occurence of that location
+        if match_step.as_block.location in location_to_predicates:
+            where_block = Filter(
+                _predicate_list_to_where_block(
+                    location_to_predicates[match_step.as_block.location]
+                )
+            )
+            # Delete the location entry. No further filters needed for this location.
+            del location_to_predicates[match_step.as_block.location]
+        else:
+            where_block = None
+        new_match_step = MatchStep(
+            root_block=match_step.root_block,
+            coerce_type_block=match_step.coerce_type_block,
+            where_block=where_block,
+            as_block=match_step.as_block
+        )
+        new_match_traversal.append(new_match_step)
+    return new_match_traversal
+
+
+def collect_filters_to_first_location_instance(compound_match_query):
+    """Collate all filters for a particular location to the first instance of the location."""
+    new_match_queries = []
+    # Each MatchQuery is processed independently
+    for match_query in compound_match_query.match_queries:
+        location_to_predicates = _construct_location_to_predicate_list(match_query)
+
+        new_match_traversals = []
+        for match_traversal in match_query.match_traversals:
+            new_match_traversal = _collect_filters_to_first_location_in_match_traversal(
+                match_traversal, location_to_predicates)
+            new_match_traversals.append(new_match_traversal)
+        new_match_queries.append(
+            MatchQuery(
+                match_traversals=new_match_traversals,
+                folds=match_query.folds,
+                output_block=match_query.output_block
+            )
+        )
+
+    return CompoundMatchQuery(match_queries=new_match_queries)
+
+
+def _update_tagged_expression(expression, current_locations):
+    """Replace non-existent tag Expressons with True, and simplify the result."""
```

Comment:
Full docstring here as well, please. This logic is far from obvious, and I bet nobody else would get how this function fits into the big picture here.

A few other specific concerns:
- There is no such thing as a `tag expression`, so this is confusing. All tag info has been stripped by the time we have IR, and tags are now `ContextField` expressions.
- This function doesn't seem to take any arguments that would specify which "tags" (presumably ContextFields?) do or don't exist, so it's not clear how it works.